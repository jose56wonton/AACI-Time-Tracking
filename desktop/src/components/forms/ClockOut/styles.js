const styles = theme => ({
  hero: {
    height: '100vh',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center'
  },
  textField: { width: '100%' },
  heroContent: {
    width: '100%',
    maxWidth: '1000px'
  },
  error: { marginTop: '16px' },
  spaceBetween: {
    justifyContent: 'space-between'
  },
  flexEnd: {
    alignItems: 'flex-end',
    justifyContent: 'flex-end'
  },
  button: { marginLeft: '10px' },

  verticalCenter: {
    display: 'flex',
    alignItems: 'center'
  },

  formHeader: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    marginRight: theme.spacing.unit
  },
  formHeaderIcon: {
    fontSize: '32px',
    margin: '0 10px'
  },

  formBody: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  formElement: {
    marginRight: theme.spacing.unit
  },

  formFooter: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between'
  },
  lunchBox: {
    width: '200px',
  }
});

export default styles;
