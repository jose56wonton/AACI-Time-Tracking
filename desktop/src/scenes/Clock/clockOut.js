import React, { Component } from 'react';
import PropTypes from 'prop-types';

import ClockOutContainer from 'containers/Clock/clockOutContainer';

class ClockOut extends Component {
  render() {
    return <ClockOutContainer type={this.props.type} />;
  }
}

ClockOut.propTypes = {
  type: PropTypes.string.isRequired
};

export default ClockOut;
