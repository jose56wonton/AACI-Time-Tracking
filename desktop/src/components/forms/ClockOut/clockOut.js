import React from 'react';
import PropTypes from 'prop-types';

import cx from 'classnames';
import { withStyles } from '@material-ui/core/styles';
import { Button, Grid, Typography, IconButton, MenuItem, SvgIcon } from '@material-ui/core';
import { AccessTime as TimeIcon, Timer as DurationIcon, Today as DateIcon } from '@material-ui/icons';
import { Field, FieldArray } from 'formik';
import { Close } from '@material-ui/icons';
import styles from './styles';
import TextField from 'components/inputs/TextField';
import Select from 'components/inputs/Select';
import Time from 'components/inputs/Time';

const ClockOutForm = props => {
  const { classes, isSubmitting, handleSubmit, shift, values, projects, projectTasks, cancel, errors } = props;
  let errorGeneralMessage;
  if (errors.activities && typeof errors.activities === 'string'){
    errorGeneralMessage = errors.activities;
  }else if (errors.lunch && typeof errors.lunch === 'string'){
    errorGeneralMessage = errors.lunch;
  }
  
  console.log(values.activities,errors.activities);
  return (
    <div className={classes.hero}>
      <div className={classes.heroContent}>
        <form onSubmit={handleSubmit}>
          <Grid container spacing={24}>
            <Grid item xs={12} className={classes.formHeader}>
              <Typography variant="display2">Clock Out</Typography>

              <div className={classes.formHeader}>
                <SvgIcon className={classes.formHeaderIcon} color="action">
                  <DateIcon />
                </SvgIcon>
                <Typography variant="headline">{shift.date}</Typography>
              </div>

              <div className={classes.formHeader}>
                <SvgIcon className={classes.formHeaderIcon} color="action">
                  <TimeIcon />
                </SvgIcon>
                <Typography variant="headline">
                  {shift.in} - {shift.out}
                </Typography>
              </div>

              <div className={classes.formHeader}>
                <SvgIcon className={classes.formHeaderIcon} color="action">
                  <DurationIcon />
                </SvgIcon>
                <Typography variant="headline">
                  Length: {shift.length}
                </Typography>
              </div>
            </Grid>

            <Grid item xs={12}>
              <FieldArray
                name="activities"
                render={arrayHelpers => {
                  return (
                    <div>
                      {values.activities &&
                        values.activities.map((activity, index) => {
                          return (
                            <div
                              key={index}
                              className={cx(
                                classes.card,
                                classes.verticalCenterBox,
                              )}
                            >
                              <div className={classes.formBody}>
                                <Field
                                  name={`activities.${index}.projectId`}
                                  component={Select}
                                  items={projects}
                                  fullWidth
                                  label="Project"
                                />
                                <Field
                                  name={`activities.${index}.projectTaskId`}
                                  component={Select}
                                  fullWidth
                                  label="Task"
                                >
                                  {
                                    projectTasks // This code iterates the projectTask 
                                      .filter(projectTask => {
                                        return activity.projectId === projectTask.projectId; // filters based on project selected
                                      })
                                      .map((projectTask, i) => { // maps those elements
                                        return (
                                          <MenuItem
                                            key={i}
                                            id="projectTaskId"
                                            value={projectTask.id}
                                          >
                                            {projectTask.task.name}
                                          </MenuItem>
                                        );
                                      })
                                  }
                                </Field>


                                <Field
                                  name={`activities.${index}.length`}
                                  component={Time}
                                  fullWidth
                                />
                                <Field
                                  name={`activities.${index}.description`}
                                  label="Description"
                                  component={TextField}
                                />
                                <div className={classes.verticalCenter}>
                                  <IconButton
                                    type="button"
                                    color="secondary"
                                    className={classes.iconButton}
                                    onClick={() => arrayHelpers.remove(index)}
                                  >
                                    <Close />
                                  </IconButton>
                                </div>
                              </div>
                            </div>
                          );
                        })}

                      <Grid item xs={12} className={classes.formFooter}>
                        <div className={classes.lunchBox}>
                          <Field
                            name={`lunch`}
                            label1="Lunch"
                            label2=" "
                            fullWidth
                            margin="none"
                            component={Time}
                          />
                        </div>
                        <Button
                          color="primary"
                          variant="contained"
                          onClick={() =>
                            arrayHelpers.push({
                              projectId: Object.keys(projects)[0],
                              projectTaskId: -1,
                              length: 0,
                              description: ''
                            })
                          }
                        >
                          Add Activity
                        </Button>
                      </Grid>

                      <Grid item xs={12} className={classes.formFooter}>

                        <Typography variant="body1" margin="none" className={classes.error}>
                          {errorGeneralMessage}
                        </Typography>

                        <div>
                          <Button
                            type="submit"
                            color="primary"
                            disabled={isSubmitting || Object.keys(errors).length !== 0}
                            variant="contained"
                            className={classes.button}
                          >
                            Clock Out
                          </Button>

                          <Button
                            onClick={cancel}
                            color="secondary"
                            variant="text"
                            className={classes.button}
                          >
                            Cancel
                          </Button>
                        </div>
                      </Grid>
                    </div>
                  );
                }}
              />
            </Grid>
          </Grid>
        </form>
      </div>
    </div>
  );
};

ClockOutForm.propTypes = {
  classes: PropTypes.object.isRequired,
  isSubmitting: PropTypes.bool.isRequired,
  values: PropTypes.object.isRequired,
  setFieldValue: PropTypes.func,
  handleSubmit: PropTypes.func,
  shift: PropTypes.object,
  projects: PropTypes.array,
  projectTasks: PropTypes.array,
  cancel: PropTypes.func
};

export default withStyles(styles)(ClockOutForm);
