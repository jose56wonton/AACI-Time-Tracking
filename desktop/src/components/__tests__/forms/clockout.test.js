import React from 'react';
import { shallow } from 'enzyme';
import { FieldArray } from 'formik';

import { Clockout } from 'components/forms/ClockOut/clockOut';
import ClockoutHOC from 'components/forms/ClockOut';

const props =  {  
  classes: {},
  label: 'label',
  isSubmitting: true,
  resetForm: jest.fn(),
  initialValues: {},
  values: {
    activities: [{ projectId:1 },{ projectId:0 }]
  },
  setFieldValue: jest.fn(),
  handleSubmit: jest.fn(),
  shift: {},
  projects: [{ id: 1 },{ id: 0 }],
  projectTasks: [{ projectId:1,task: { name: 'name' } },{ projectId: 0,task: { name: 'name' } }],
  cancel: jest.fn(),
  timeLeft: 0,
  generalError: 'general error',
  errors: {}
};
const renderProps = { 
  remove: jest.fn(),
  push: jest.fn()
};

const setup = overRides => {  
  return shallow(<Clockout {...props} {...overRides}/>);    
};

const setupHOC = overRides => {
  return shallow(<ClockoutHOC {...props} {...overRides}/>);
};

const setupWithRender = overRides => {
  const wrapper = setup();  
  const Render = wrapper.find(FieldArray).first().prop('render');  
  return shallow(<Render {...renderProps} {...overRides}/>);
};

describe('Clockout Component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('should render correctly', () => {
    setup();   
  });
  it('should render correctly withStyles', () => {
    setupHOC();
  });
  it('should call render of acitvities fieldarray', () => {
    setupWithRender();
  });
  it('should remove an activity when a remove-activity-{1}', () => {
    const wrapper = setupWithRender();
    expect(renderProps.remove).toHaveBeenCalledTimes(0);
    wrapper.find('#remove-activity-1').first().simulate('click');
    expect(renderProps.remove).toHaveBeenCalledTimes(1);
    expect(renderProps.remove).toHaveBeenCalledWith(1);
  });
  it('should remove an activity when a remove-activity-{1}', () => {
    const wrapper = setupWithRender();
    expect(renderProps.push).toHaveBeenCalledTimes(0);
    wrapper.find('#add-activity').first().simulate('click');
    expect(renderProps.push).toHaveBeenCalledTimes(1);
  });
});