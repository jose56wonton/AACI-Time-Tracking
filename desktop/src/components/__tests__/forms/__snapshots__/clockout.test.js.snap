// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clockout Component should render correctly 1`] = `
<div>
  <div>
    <form
      onSubmit={[Function]}
    >
      <WithStyles(Grid)
        container={true}
        spacing={24}
      >
        <WithStyles(Grid)
          item={true}
          xs={12}
        >
          <WithStyles(Typography)
            variant="h3"
          >
            Clock Out
          </WithStyles(Typography)>
          <div>
            <WithStyles(SvgIcon)
              color="action"
            >
              <pure(TodayIcon) />
            </WithStyles(SvgIcon)>
            <WithStyles(Typography)
              variant="h5"
            />
          </div>
          <div>
            <WithStyles(SvgIcon)
              color="action"
            >
              <pure(AccessTimeIcon) />
            </WithStyles(SvgIcon)>
            <WithStyles(Typography)
              variant="h5"
            >
               - 
            </WithStyles(Typography)>
          </div>
          <div>
            <WithStyles(SvgIcon)
              color="action"
            >
              <pure(TimerIcon) />
            </WithStyles(SvgIcon)>
            <WithStyles(Typography)
              variant="h5"
            >
              Length: 
            </WithStyles(Typography)>
          </div>
        </WithStyles(Grid)>
        <WithStyles(Grid)
          item={true}
          xs={12}
        >
          <C
            id="activities"
            name="activities"
            render={[Function]}
          />
        </WithStyles(Grid)>
      </WithStyles(Grid)>
    </form>
  </div>
</div>
`;

exports[`Clockout Component should render correctly withStyles 1`] = `
<Component
  cancel={[Function]}
  classes={
    Object {
      "button": "Component-button-7",
      "error": "Component-error-4",
      "flexEnd": "Component-flexEnd-6",
      "formBody": "Component-formBody-11",
      "formElement": "Component-formElement-12",
      "formFooter": "Component-formFooter-13",
      "formHeader": "Component-formHeader-9",
      "formHeaderIcon": "Component-formHeaderIcon-10",
      "hero": "Component-hero-1",
      "heroContent": "Component-heroContent-3",
      "lunchBox": "Component-lunchBox-14",
      "spaceBetween": "Component-spaceBetween-5",
      "textField": "Component-textField-2",
      "verticalCenter": "Component-verticalCenter-8",
    }
  }
  errors={Object {}}
  generalError="general error"
  handleSubmit={[Function]}
  initialValues={Object {}}
  isSubmitting={true}
  label="label"
  projectTasks={
    Array [
      Object {
        "projectId": 1,
        "task": Object {
          "name": "name",
        },
      },
      Object {
        "projectId": 0,
        "task": Object {
          "name": "name",
        },
      },
    ]
  }
  projects={
    Array [
      Object {
        "id": 1,
      },
      Object {
        "id": 0,
      },
    ]
  }
  resetForm={[Function]}
  setFieldValue={[Function]}
  shift={Object {}}
  timeLeft={0}
  values={
    Object {
      "activities": Array [
        Object {
          "projectId": 1,
        },
        Object {
          "projectId": 0,
        },
      ],
    }
  }
/>
`;
